### PR TITLE
Ed/view sync task misc

### PR DESCRIPTION
**Pull request by elliedavidson**
This PR: 
* Addresses comments from my last PR
* Adds ViewSyncCertSend/Recv and ViewSyncVoteSend/Recv events
  * Adds handling for the above in the network task
* Adds some extra utility functions on view sync types

Still todo
* Initialize view sync task in mod.rs

### CODE REVIEW

File: task-impls/src/view_sync.rs [hunk 1]
```diff
@@ -386,12 +393,18 @@ where
...
+                                    .register_event_stream(relay_state.event_stream.clone(), filter)
+                                    .await
+                                    .register_registry(&mut self.registry.clone())
                                     .await
                                     .register_state(relay_state)
                                     .register_event_handler(relay_handle_event);
+
+                            // TODO ED For now we will not await these futures, in the future we can await them only in the case of shutdown
+                            let _view_sync_relay_task =
+                                ViewSyncRelayTaskStateTypes::build(builder).launch();
```
Comment (bfish713) [inline task-impls/src/view_sync.rs:407]: I think this is where we need to async_spawn and await or the future won't start running

Comment (elliedavidson) [reply task-impls/src/view_sync.rs:407]: yes, you're right!  I'll add that in! Slipped my mind

**Review by bfish713 — APPROVED** [review]
just one small comment on vote collection, otherwise looks good!

File: task-impls/src/view_sync.rs [hunk 2]
```diff
@@ -317,15 +320,20 @@ where
...
                                 .await
+                                .register_registry(&mut self.registry.clone())
+                                .await
                                 .register_state(replica_state)
                                 .register_event_handler(replica_handle_event);
-                            // TODO ED Launch task
+
+                            let _view_sync_replica_task =
+                            async_spawn(
+                                ViewSyncReplicaTaskStateTypes::build(builder).launch());
```
Comment (DieracDelta) [inline task-impls/src/view_sync.rs:336]:
```suggestion
                            async_spawn(
                                ViewSyncReplicaTaskStateTypes::build(builder).launch()).await;
```
`launch` returns a `BoxFuture` that needs to be awaited in order to be executed

File: task-impls/src/consensus.rs [hunk 3]
```diff
@@ -563,18 +563,20 @@ where
...
                 }
             }
-            SequencingHotShotEvent::ViewChange(_) => nll_todo(),
-            SequencingHotShotEvent::ViewSyncMessageRecv(_) => {
+
+            SequencingHotShotEvent::ViewChange(new_view) => {
                 // update the view in state to the one in the message
-                // TODO ED This info should come in the form of a ViewChange message, update
-                nll_todo()
+                if self.cur_view < TYPES::Time::new(*new_view) {
```
[the remark this answers is not in the document]
Comment (bfish713) [reply task-impls/src/consensus.rs:577]: I think it's ok for now, we probably should cancel some tasks, but they will get cancelled either way when other messages come in.  I need to think it through a little

File: task-impls/src/view_sync.rs [hunk 4]
```diff
@@ -386,12 +393,18 @@ where
...
+                                    .register_event_stream(relay_state.event_stream.clone(), filter)
+                                    .await
+                                    .register_registry(&mut self.registry.clone())
                                     .await
                                     .register_state(relay_state)
                                     .register_event_handler(relay_handle_event);
+
+                            // TODO ED For now we will not await these futures, in the future we can await them only in the case of shutdown
+                            let _view_sync_relay_task =
+                            async_spawn(ViewSyncRelayTaskStateTypes::build(builder).launch());
```
Comment (DieracDelta) [inline task-impls/src/view_sync.rs:407]: Is the idea to conditionally await this?

**Review by DieracDelta — APPROVED** [review]

**Pull request merged**